### PR TITLE
Range-separated density fitting for PBC calculations

### PR DESCRIPTION
**Pull request by hongzhouye**
This PR adds the implementation of the range-separated gaussian density fitting (RSGDF) developed by @tberkel and me as detailed in the following two manuscripts:
1. Ye, and Berkelbach _J. Chem. Phys._ **154**, 131104 (2021) DOI: 10.1063/5.0046617
2. Ye, and Berkelbach _J. Chem. Phys._ **155**, 124106 (2021) DOI: 10.1063/5.0064151

It is implemented as a derived class of `pbc.df.df.GDF` and can be used in almost the same way as `GDF`. E.g.,
```Python
mf = scf.KRHF(cell, kpts).rs_density_fit()
```

We did a decent amount of tests that cover (huge thanks to @btglau, @verena-neufeld, @zhutianyu1991, @Majokux, and @cnyeh for running tests and providing useful feedbacks on preliminary versions of this module)
- different materials and basis sets
- different levels of theory including SCF, MP2, CCSD, EOM-CCSD, 
- k-point sampling vs gamma point with big supercell

against the current `GDF` in PySCF. The agreement between RSGDF and GDF is in general perfect, with a typical difference in energy 10^-8 Eh per cell when using the default `cell.precision`. The speedup of RSGDF warm-up over GDF is typically 10 ~ 100x when measured in terms of CPU time or 5 ~ 10x in wall time depending on the system/basis sets and number of CPU cores being used. The smaller speedup in wall time is due to the I/O bound affecting RSGDF more (because it's faster).

### CODE REVIEW

File: pyscf/pbc/df/rsdf.py [hunk 1]
```diff
@@ -0,0 +1,703 @@
...
+    else:
+        bvk_kmesh = None
+
+    # The ideal way to hold the temporary integrals is to store them in the
+    # cderi_file and overwrite them inplace in the second pass.  The current
+    # HDF5 library does not have an efficient way to manage free space in
+    # overwriting.  It often leads to the cderi_file ~2 times larger than the
+    # necessary size.  For now, dumping the DF integral intermediates to a
+    # separated temporary file can avoid this issue.  The DF intermediates may
+    # be terribly huge. The temporary file should be placed in the same disk
```
Comment (verena-neufeld) [inline pyscf/pbc/df/rsdf.py:141]: Why should it be placed on the same disk?

File: pyscf/pbc/df/rsdf.py [hunk 2]
```diff
@@ -0,0 +1,703 @@
...
+    '''
+
+    # class methods defined outside the class
+    _make_j3c = _make_j3c
+
+    def __init__(self, cell, kpts=np.zeros((1,3))):
+        if cell.dimension < 3:
+            raise NotImplementedError("RSGDF for low-dimensional systems are not available yet. We recommend using cell.dimension=3 with large vacuum.")
+
+        # if True and kpts are gamma-inclusive, RSDF will use the bvk cell trick for computing both j3c_SR and j3c_LR. If kpts are not gamma-inclusive, this attribute will be ignored.
```
Comment (verena-neufeld) [inline pyscf/pbc/df/rsdf.py:468]: This is quite a long line

File: pyscf/pbc/df/rsdf.py [hunk 3]
```diff
@@ -0,0 +1,703 @@
...
+        self.use_bvk = True
+
+        # precision for real-space lattice sum (R) and reciprocal-space Fourier transform (G).
+        self.precision_R = cell.precision * 1e-2
+        self.precision_G = cell.precision
+
+        # One of {omega, npw_max} must be provided, and the other will be deduced automatically from it. The priority when both are given is omega > npw_max.
```
Comment (verena-neufeld) [inline pyscf/pbc/df/rsdf.py:475]: This is quite a long line

File: pyscf/pbc/df/rsdf.py [hunk 4]
```diff
@@ -0,0 +1,703 @@
...
+        # Once omega is determined, mesh_compact is determined for (L|g^lr|pq) to achieve given accuracy.
+        self.npw_max = 350
+        self._omega_min = 0.1
+        self.omega = None
+        self.ke_cutoff = None
+        self.mesh_compact = None
+
+        # omega and mesh for j2c. Since j2c is to be inverted, it is desirable to use (1) a "fixed" omega for reproducibility, and (2) a higher precision to minimize any round-off error caused by inversion.
+        # The extra computational cost due to the higher precision is negligible compared to the j3c build.
+        # FOR EXPERTS: if you want omega_j2c to be the same as the omega used for j3c build, set omega_j2c to be any negative number.
```
Comment (verena-neufeld) [inline pyscf/pbc/df/rsdf.py:487]: is this safe? would it make sense to allow a new variable the user can pass?

File: pyscf/pbc/df/rsdf_helper.py [hunk 5]
```diff
@@ -0,0 +1,1229 @@
...
+""" Prescreening 2c
+"""
+def Gamma(s, x):
+    return gammaincc(s,x) * gamma(s)
+def get_multipole(l, alp):
+    return 0.5*np.pi * (2*l+1)**0.5 / alp**(l+1.5)
+def get_2c2e_Rcut(bas_lst, cellvol, omega, precision, Rprec=1.,
+                  lmp=True, lasympt=True,
+                  eta_correct=True, R_correct=False, vol_correct=False):
+    r""" Given a list of pgto by "bas_lst", determine the cutoff radii for j2c lat sum s.t. the truncation error drops below "precision". j2c is estimated as
```
Comment (verena-neufeld) [inline pyscf/pbc/df/rsdf_helper.py:202]: This line is quite long

File: pyscf/pbc/df/rsdf_helper.py [hunk 6]
```diff
@@ -0,0 +1,1229 @@
...
+    return I
+def get_schwartz_data(bas_lst, omega, dijs_lst=None, keep1ctr=True, safe=True):
+    """
+        if dijs_lst is None:
+            "2c"-mode:  Q = 2-norm[(a|a)]^(1/2)
+        else:
+            "4c"-mode:  Q = 2-norm[(ab|ab)]^(1/2)
+    """
+    def get1ctr(bas_lst):
+        """ For a shell consists of multiple contracted GTOs, keep only the one with the greatest weight on the most diffuse primitive GTOs (since others are likely core orbitals).
```
Comment (verena-neufeld) [inline pyscf/pbc/df/rsdf_helper.py:322]: This is quite a long line

File: pyscf/pbc/df/rsdf_helper.py [hunk 7]
```diff
@@ -0,0 +1,1229 @@
...
+                shls_slice = (i,i+1,j_,j_+1,i,i+1,j_,j_+1)
+                dijs = dijs_lst[ij]
+                Qs[ij] = [compute1_(mol, dij, intor, shls_slice, omega, safe)
+                          for dij in dijs]
+                ij += 1
+
+    return Qs
+def get_schwartz_dcut(bas_lst, cellvol, omega, precision, r0=None, safe=True,
+                      vol_correct=False):
+    """ Given a list of basis, determine cutoff radius for the Schwartz Q between each unique shell pair to drop below "precision". The Schwartz Q is define:
```
Comment (verena-neufeld) [inline pyscf/pbc/df/rsdf_helper.py:376]: This is quite a long line

File: pyscf/pbc/df/rsdf_helper.py [hunk 8]
```diff
@@ -0,0 +1,1229 @@
...
+
+    intor = "int2e"
+    def estimate1(ish,jsh,R0,R1):
+        shls_slice = (ish,ish+1,nbas+jsh,nbas+jsh+1,
+                      ish,ish+1,nbas+jsh,nbas+jsh+1)
+        prec0 = precision * min(etas[ish,jsh],1.)
+        def fcheck(R):
+            mol._env[mol._atm[1,mol_gto.PTR_COORD]] = R
+            I = get_norm(fintor_sreri(mol, intor, shls_slice, omega, safe))**0.5
+# >>>>>>> debug block
```
Comment (verena-neufeld) [inline pyscf/pbc/df/rsdf_helper.py:403]: What do you mean with debug block?

File: pyscf/pbc/df/rsdf_helper.py [hunk 9]
```diff
@@ -0,0 +1,1229 @@
...
+                lmin = 0
+                L_{li,lj}^{l} = \sum'_{m=-l}^{l} comb(li,mi) * comb(lj,mj) * di^(li-mi) * dj^(lj-mj)
+                where
+                    mi = (l+m)/2
+                    mj = (l-m)/2
+                    di = -ej/eij * (dij + extij)
+                    dj = ei/eij * (dij + extij)
+                where "extij" is the extent of orbital pair ij.
+
+        Similar to :func:`get_2c2e_Rcut`, the estimator is multiplied by factor of eta and/or 1/R if "eta_correct" and/or "R_correct" are set to True.
```
Comment (verena-neufeld) [inline pyscf/pbc/df/rsdf_helper.py:466]: This is quite a long line

File: pyscf/pbc/df/rsdf_jk.py [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+# limitations under the License.
+#
+# Author: Hong-Zhou Ye <hzyechem@gmail.com>
+#
+
+import numpy
+import copy
+
+def density_fit(mf, auxbasis=None, mesh=None, with_df=None):
+    '''Generte density-fitting SCF object
```
Comment (verena-neufeld) [inline pyscf/pbc/df/rsdf_jk.py:23]: Generate

File: pyscf/pbc/df/test/test_rsdf.py [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+
+def tearDownModule():
+    global cell
+    del cell
+
+class KnownValues(unittest.TestCase):
+    def test_h2_gamma(self):
+        mf = pscf.KRHF(cell).rs_density_fit()
+        mf.kernel()
+        self.assertAlmostEqual(mf.e_tot, -1.0430635251, 6)
```
Comment (verena-neufeld) [inline pyscf/pbc/df/test/test_rsdf.py:43]: Is `6` the accuracy you'd expect with the default RSGDF settings?

**Review by verena-neufeld — COMMENTED** [review]
Hi Hong-Zhou! Super cool stuff, I'll continue testing this :D. For now, I have started with a very superficial review of some super minor comments. Feel free to ignore as many of these as you like. I think some tests are currently failing here due to flake8.

**Comment by sunqm** [comment]
In nr3c_k and other `fill` functions, the short range Coulomb integrals are computed as `full_integrals - long_range_integrals` . They can be computed directly using the releases of libcint 4.4.4 and newer. Could you merge the master and test whether it gives different output and changes performance?

File: pyscf/pbc/df/rsdf.py [hunk 12]
```diff
@@ -0,0 +1,703 @@
...
+[2] For the SR lattice sum integral screening:
+    Hong-Zhou Ye and Timothy C. Berkelbach, arXiv:2107.09704.
+
+In RSGDF, the two-center and three-center Coulomb integrals are calculated in two pars:
+    j2c = j2c_SR(omega) + j2c_LR(omega)
+    j3c = j3c_SR(omega) + j3c_LR(omega)
+where the SR and LR integrals correpond to using the following potentials
+    g_SR(r_12;omega) = erfc(omega * r_12) / r_12
+    g_LR(r_12;omega) = erf(omega * r_12) / r_12
+The SR integrals are evaluated in real space using a lattice summation, while the LR integrals are evaluated in reciprocal space with a plane wave basis.
```
[the remark this answers is not in the document]
Comment (hongzhouye) [reply pyscf/pbc/df/rsdf.py]: fixed.

File: pyscf/pbc/df/rsdf.py (hunk 1 above)
Comment (hongzhouye) [reply pyscf/pbc/df/rsdf.py:150]: these lines were copied from GDF

File: pyscf/pbc/df/rsdf.py [hunk 13]
```diff
@@ -0,0 +1,703 @@
...
+    def cholesky_decomposed_metric(uniq_kptji_id):
+        j2c = np.asarray(fswap['j2c/%d'%uniq_kptji_id])
+        j2c_negative = None
+        try:
+            if mydf.j2c_eig_always:
+                raise scipy.linalg.LinAlgError
+            j2c = scipy.linalg.cholesky(j2c, lower=True)
+            j2ctag = 'CD'
+        except scipy.linalg.LinAlgError:
+            #msg =('===================================\n'
```
[the remark this answers is not in the document]
Comment (hongzhouye) [reply pyscf/pbc/df/rsdf.py:231]: these lines are copied from GDF so I kept them unchanged.

File: pyscf/pbc/df/rsdf.py (hunk 2 above)
Comment (hongzhouye) [reply pyscf/pbc/df/rsdf.py]: fixed.

File: pyscf/pbc/df/rsdf.py (hunk 3 above)
Comment (hongzhouye) [reply pyscf/pbc/df/rsdf.py]: fixed.

File: pyscf/pbc/df/rsdf_helper.py (hunk 7 above)
Comment (hongzhouye) [reply pyscf/pbc/df/rsdf_helper.py]: fixed.

File: pyscf/pbc/df/rsdf_helper.py (hunk 8 above)
Comment (hongzhouye) [reply pyscf/pbc/df/rsdf_helper.py:460]: for experts :D

File: pyscf/pbc/df/rsdf_helper.py (hunk 9 above)
Comment (hongzhouye) [reply pyscf/pbc/df/rsdf_helper.py]: fixed.

File: pyscf/pbc/df/rsdf_jk.py (hunk 10 above)
Comment (hongzhouye) [reply pyscf/pbc/df/rsdf_jk.py:23]: Copied from `df_jk.py` so I will leave this feature typo for now.

File: pyscf/pbc/df/test/test_rsdf.py (hunk 11 above)
Comment (hongzhouye) [reply pyscf/pbc/df/test/test_rsdf.py]: Changed to 7. RSDF computes integrals to be `cell.precision` which by default is 1e-8. It's not clear exactly how that is translated to the precision in SCF energy but a bit larger than that is expected in general.

**Comment by hongzhouye** [comment]
@sunqm Hi Qiming, as we discussed last time, I
1. fixed the get_lattice_Ls bug in [this commit](https://github.com/pyscf/pyscf/pull/1034/commits/84106178bdeb6a8807a21e91478a8d45bd443497).
2. added tests in [this commit](https://github.com/pyscf/pyscf/pull/1034/commits/213cd38faddc938791ade775bcf686219da96063). They are mostly taken from the corresponding tests for GDF, with known values being modified slightly to account for the small difference between RSDF and GDF (1e-6 ~ 1e-5 for most matrix elements).
3. added examples in [this commit](https://github.com/pyscf/pyscf/pull/1034/commits/f20779b9a78e3f9bdf74a22b4e226115a19e62b2).

I hope these changes make the PR ready for a final review.

File: pyscf/pbc/df/rsdf_helper.py [hunk 14]
```diff
@@ -0,0 +1,1297 @@
+#!/usr/bin/env python
```
Comment (sunqm) [inline pyscf/pbc/df/rsdf_helper.py:1]: Many functions in this file are needed only by this module. Can you prefix them with `_` to indicate they are private functions? For functions which may be used by other modules, please add docs for all arguments.

File: pyscf/pbc/df/rsdf_helper.py [hunk 15]
```diff
@@ -0,0 +1,1297 @@
...
+        lpmax = min(l,2*l1-l)
+        for lp in range(lpmin,lpmax+1,2):
+            l1p = (l+lp) // 2
+            l2p = (l-lp) // 2
+            cl += d1**(l1-l1p)*d2**(l2-l2p) * comb(l1,l1p) * comb(l2,l2p)
+        cbins[l] = cl
+    return cbins
+def get_3c2e_Rcuts_for_d(mol, auxmol, ish, jsh, dij, cellvol, omega, precision,
+                         fac_type, Qij, Rprec=1,
+                         vol_correct=False, eta_correct=True, R_correct=True):
```
Comment (sunqm) [inline pyscf/pbc/df/rsdf_helper.py:496]: What are fac_type, Qij, Rprec, vol_correct, eta_correct, R_correct? Please add docs for these arguments.

File: pyscf/pbc/df/rsdf_helper.py [hunk 16]
```diff
@@ -0,0 +1,1297 @@
...
+    expkL = np.asarray(np.exp(1j*np.dot(kpts_lst, Ls.T)), order='C')
+    drv = libpbc.PBCsr2c_k_drv
+
+    drv(fintor, fill, out.ctypes.data_as(ctypes.c_void_p),
+        ctypes.c_int(nkpts), ctypes.c_int(comp), ctypes.c_int(len(Ls)),
+        Ls.ctypes.data_as(ctypes.c_void_p),
+        expkL.ctypes.data_as(ctypes.c_void_p),
+        (ctypes.c_int*4)(i0, i1, j0, j1),
+        ao_loc.ctypes.data_as(ctypes.c_void_p), cintopt,
+        refuniqshl_map.ctypes.data_as(ctypes.c_void_p),
```
Comment (sunqm) [inline pyscf/pbc/df/rsdf_helper.py:905]: ensure dtype of refuniqshl_map. (call np.asarray(..., dtype=) before calling the C function)

**Comment by sunqm** [comment]
`rs_density_fit` is supposed to deliver a 3-index tensor the same to the one from `density_fit`, right? When using the default parameters, which one is not tight enough that causes the small differences between the two?

File: pyscf/lib/pbc/fill_ints_sr.c [hunk 17]
```diff
@@ -0,0 +1,2002 @@
...
+ */
+
+#include <stdlib.h>
+#include <complex.h>
+#include <assert.h>
+#include "config.h"
+#include "cint.h"
+#include "vhf/fblas.h"
+#include "np_helper/np_helper.h"
+
```
[the remark this answers is not in the document]
Comment (hongzhouye) [reply pyscf/lib/pbc/fill_ints_sr.c:27]: fixed in [this commit](https://github.com/pyscf/pyscf/pull/1034/commits/669cba9c9c1315d7f2269c161bafbd01bf038cc7)

File: pyscf/pbc/df/rsdf_helper.py (hunk 14 above)
Comment (hongzhouye) [reply pyscf/pbc/df/rsdf_helper.py:1]: fixed in [this commit](https://github.com/pyscf/pyscf/pull/1034/commits/94585b16b1987beda31e273b0e5e2bc003e76f01)

File: pyscf/pbc/df/rsdf_helper.py (hunk 15 above)
Comment (hongzhouye) [reply pyscf/pbc/df/rsdf_helper.py]: fixed in [this commit](https://github.com/pyscf/pyscf/pull/1034/commits/94585b16b1987beda31e273b0e5e2bc003e76f01)

File: pyscf/pbc/df/rsdf_helper.py [hunk 18]
```diff
@@ -0,0 +1,1297 @@
...
+
+    return Rcuts
+def get_3c2e_Rcuts(bas_lst_or_mol, auxbas_lst_or_auxmol,
+                   dijs_lst, cellvol, omega, precision,
+                   fac_type, Qijs_lst, Rprec=1,
+                   eta_correct=True, R_correct=True, vol_correct=False):
+    """ Given a list of basis ("bas_lst") and auxiliary basis ("auxbas_lst"),
+    determine the cutoff radius for
+        2-norm( (k|v_SR(omega)|ij) ) < precision
+    where i and j shls are separated by d specified by "dijs_lst".
```
[the remark this answers is not in the document]
Comment (hongzhouye) [reply pyscf/pbc/df/rsdf_helper.py:658]: fixed in [this commit](https://github.com/pyscf/pyscf/pull/1034/commits/94585b16b1987beda31e273b0e5e2bc003e76f01)

File: pyscf/pbc/df/rsdf_helper.py (hunk 16 above)
Comment (hongzhouye) [reply pyscf/pbc/df/rsdf_helper.py:929]: the dtype of `refuniqshl_map` is already set to numpy.int32 in the function generating it `_get_refuniq_map`. See L171.

Do I need to add another `np.asarray(..., dtype=)` right before calling the C function in `intor_j2c`?

File: pyscf/pbc/df/rsdf_helper.py [hunk 19]
```diff
@@ -0,0 +1,1297 @@
...
+                          nbas+shls_slice[2], nbas+shls_slice[3],
+                          nbas*2+shls_slice[4], nbas*2+shls_slice[5])
+            drv(getattr(libpbc, intor), getattr(libpbc, fill),
+                out.ctypes.data_as(ctypes.c_void_p),
+                ctypes.c_int(comp), ctypes.c_int(nimgs),
+                Ls.ctypes.data_as(ctypes.c_void_p),
+                (ctypes.c_int*6)(*shls_slice),
+                ao_loc.ctypes.data_as(ctypes.c_void_p),
+                cintopt,
+                shlpr_mask.ctypes.data_as(ctypes.c_void_p),  # shlpr_mask
```
[the remark this answers is not in the document]
Comment (hongzhouye) [reply pyscf/pbc/df/rsdf_helper.py:1244]: `shlpr_mask` is set to `numpy.int8`; see L1053
`refuniqshl_map` and `auxuniqshl_map` are set to `numpy.int32` in the function generating them;  see L171
`dijs_loc` is set to `numpy.int32` when generated; see L1007
Other arrays are all double (or like `atm`, `bas` etc directly taken from `cell`).

File: pyscf/pbc/df/rsdf_jk.py (hunk 10 above)
Comment (hongzhouye) [reply pyscf/pbc/df/rsdf_jk.py]: fixed

**Comment by hongzhouye** [comment]
@sunqm I *think* the main part of the small difference comes from using a different auxiliary basis. RSDF uses the default while GDF re-normalizes the fitting functions.

**Review by sunqm — APPROVED** [review]

**Pull request merged**